main.py

```diff
@@ -92,54 +97,136 @@ def test_match_code_rule(
     matches = json.loads(result.stdout.strip())
     if not matches:
         raise ValueError("No matches found for the given code and rule. Try adding `stopBy: end` to your inside/has rule.")
-    return matches
+    return matches  # type: ignore[no-any-return]
 
 @mcp.tool()
 def find_code(
     project_folder: str = Field(description="The absolute path to the project folder. It must be absolute path."),
     pattern: str = Field(description="The ast-grep pattern to search for. Note, the pattern must have valid AST structure."),
     language: str = Field(description="The language of the query", default=""),
-) -> List[dict[str, Any]]:
+    max_results: Optional[int] = Field(default=None, description="Maximum results to return"),
+    output_format: str = Field(default="text", description="'text' or 'json'"),
+) -> str | List[dict[str, Any]]:
     """
     Find code in a project folder that matches the given ast-grep pattern.
     Pattern is good for simple and single-AST node result.
     For more complex usage, please use YAML by `find_code_by_rule`.
 
-    Internally calls: ast-grep run --pattern <pattern> --json <project_folder>
+    Internally calls: ast-grep run --pattern <pattern> [--json] <project_folder>
+
+    Output formats:
+    - text (default): Simple file:line:content format, ~75% fewer tokens
+    - json: Full match objects with metadata
+
+    Example usage:
+      find_code(pattern="class $NAME", max_results=20)  # Returns text format
+      find_code(pattern="class $NAME", output_format="json")  # Returns JSON with metadata
     """
-    args = ["--pattern", pattern, "--json"]
+    if output_format not in ["text", "json"]:
+        raise ValueError(f"Invalid output_format: {output_format}. Must be 'text' or 'json'.")
+
+    args = ["--pattern", pattern]
     if language:
         args.extend(["--lang", language])
-    args.append(project_folder)
-    result = run_ast_grep("run", args)
-    return json.loads(result.stdout)
+
+    if output_format == "json":
+        # JSON format - return structured data
+        result = run_ast_grep("run", args + ["--json", project_folder])
+        matches = json.loads(result.stdout.strip() or "[]")
+        # Limit results if max_results is specified
+        if max_results is not None and len(matches) > max_results:
+            matches = matches[:max_results]
+        return matches  # type: ignore[no-any-return]
+    else:
+        # Text format - return plain text output
+        result = run_ast_grep("run", args + [project_folder])
+        output = result.stdout.strip()
+        if not output:
+            output = "No matches found"
+        else:
+            # Apply max_results limit if specified
+            lines = output.split('\n')
+            non_empty_lines = [line for line in lines if line.strip()]
+            if max_results is not None and len(non_empty_lines) > max_results:
```

Comment:
using lines isn't accurate here since a match can span multiple lines. 

To make things easier I think the argument name can be changed to `max_matching_lines` or simialr

Reply:
max_results is now processed correctly for multi-line matches